_docs/documentation/admin/config.md
`txmp.docs / documentation`

# Configuration and customization

## Custom language keys

You can do a lot using this, such as changing displayed strings, customize the About dialog, create DeltaDesc annotations and so on.

Vanilla Location: `lib/i18n/lang/<id>.lang`  
Override Location: `data/i18n/<id>.lang`

```plain
# Configuration
config.app_name = Music Archive
config.app_name.title = Music Archive
```

Add your own `.lang` files at the override location to override vanilla values and define new keys.

If you don't know how, have a look at this:

- [Introducing language files](./lang.md)

## Branding

As a free application, you can change the theme color and the title of the app.

### Theme Color

Location: `internal_config/config_basic.php`

```php
define("MAIN_COLOR","E64A19"); // [默认]主颜色
define("GC_COLOR_1",MAIN_COLOR);
define("GC_COLOR_2","FFA000"); // 渐变色2
```

You can only use constant hex values, without `#`.

### App Title

Location: `internal_config/config_misc.php`

```php
"app_name" => LNG('config.app_name'),//软件名称自定义
"app_name_title" => LNG('config.app_title'),
"app_desc" => LNG('config.app_desc'), // 软件描述
```

- `app_name` the one displayed on the header bar.
- `app_name_title` the one displayed on the page title.
- `app_desc` description displayed on the page title of the homepage.

You can use `LNG()` to refer to language keys.

## Misc Config

Location: `internal_config/config_misc.php`

```php
// Timezone shift (should not be used)
"timezone" => 0,
// RemotePlay cache life, in seconds
"cache_expire" => 24*60*60*93,
// Lyrics book cache life, in seconds
"temp_expire" => 3600,

// Change rp_enabled to `true` to allow RemotePlay.
// May have risk of slowing down the service.
"rp_enabled" => false,
// Retries of RemotePlay search. Default values are recommended
"rp_search_retry" => 4,
"rp_search_retry_delay" => 0.1,
// Allow pay_play hack on RemotePlay
// - false: Paid-to-play songs will be unplayable and undownloadable.
// - true: One have full access to paid-to-play songs.
"rp_can_pay_play" => false,
"rp_pay_play_admin_only" => false,

// Allow visitor register
"can_register" => true,
// Register limit per IP
"ip_reg_limit" => 1,

// Max playlists saved by a single user
"user_playlist_quota" => 128,
// Max size of a playlist (76800 = approximately 4100 songs)
"user_playlist_limit" => 76800,
// [WIP] Max exams saved by a single user
"user_exam_quota" => 50,
// [WIP] Max recent submissions
"user_submission_capacity" => 20,

// [WIP] Max problems in a exam
"exam_problem_limit" => 512,

// Offline mode (for usage in LAN)
// - false: Use CDN to load fonts, for better speed.
// - true: Load fonts from statics, for offline accessibility.
"offline_usage" => true,

// Cache compiled lyrics
// - false: Slow loading speed
// - true: Fast loading, ~6KB extra space for each song
"compiled_cache" => true,

// Debug Mode
// * Should be false in production.
// - false: Display no errors, minimized JSON.
// - true: Display errors, prettified JSON.
"debug" => true,
// Show lyric compilation process on Debug Code screen
// * Should be false in production.
"show_comp_process" => true,
```

## Broadcast

If you want to show a notice on the music discovery page.

Location: `data/bc/bc.html`

```html
<h3>Broadcast</h3>

<p>Content here...</p>
```


changelog.txt
更新历史：

需要注意
  - [长期存在] 孤立的 jsWithPHP、cssWithPHP 格式需要进行集成。
  - [长期存在] 主页、歌曲管理和用户管理页需要筛选控件。
  - [v126c引入] 歌单构造器需要虚拟键盘按钮以提醒用户/适配移动端。
  - [长期存在] 歌单会自动移除不存在的项目而不告知用户。
  - [长期存在] 页面上的功能过于零碎。
  - [v117a引入] RemotePlay 过多的歌单加载时间很长且没有进度显示。

v126d-pre7  更新时间：2020/10/16  [当前版本]
  - [修复] 修复RemotePlay歌词中 &lt; &gt; &amp; 被编码的问题（现在不需要）。
  - [修改] 修改可打印歌词的格式
    - CIP 文字减小至 24（小四）
    - 歌词、信息块文字减小至 22（11）
    - 页面边距倍增，以适应打印并装订的需要。
  - [修复] 修复 lrc 文件中 &lt; &gt; &amp; 被二次编码的问题。
  - [修复] 修复了可能存在的问题：单曲歌词本输出中掺杂有报错导致无法打开。
  - [修复] 修复编辑器“补加L开头”功能不认识 G1 和 G2 命令的问题。
  - [修改] 修改编辑器上标签的描述文字。
  - [清理][修改] 现在歌词文档支持角色标签与其他格式标签互相嵌套。
    - 总体性能略有改善（我  信  你  个  鬼）
  - [添加] 添加用于表示“人声倒放”的符号 [R]
    - 在歌曲 [K_16644260]红昭愿 的歌词中应当被用到。
    - 规范：若对一段文字的整体添加该符号，则这段文字不应该逆向书写。该符号包含文字逆向发音的含义。
    - 文本上方以注音方式显示指向左侧的长箭头。
    - 不应该嵌套自身。
    - 在可打印文档上表达为尖括号且可以与其他格式嵌套。
    - 在 lrc 文件中表达为尖括号。

v126d-pre6  更新时间：2020/10/15  [线下开发]
  - [修复] 修复单曲 edit 和 permission 页面无法显示的问题。

v126d-pre5  更新时间：2020/10/12  [有严重问题] [线下开发]
  - [清理][修复] 重写 Router 类
    - 404 和 401 总是伴随正确的 HTTP 代码抛出。
    - 404 和 401 在非包装页面（非顶层页面）不再以网页形式呈现。
    - “权限校验错误”现在伴随“Illegal Data”状态以404页面形式呈现。
    - 整体性能略有改善。
  - [清理] 迭代哈希随机数的代码文件从 hashed_rand.function.php 改为 hashed_rand.class.php。
  - [修改][破坏] 接收相对URL的GET键名从 music 改为 _lnk。
    - 更新后必须相应修改 .htaccess 才能访问除主页外的其他网址。
  - [修复] 若 configuration.php 中未定义渐变色，则自动定义为空值。
    - 从 v126d-pre1 或更早版本直接更新不再导致错误。
  - [修改] 若歌词文件主题色未定义而定义了渐变色，则主题色为 G1 色而非默认值。
    - 移除了新文件模板中的 A 命令。

v126d-pre4  更新时间：2020/10/10  [线下开发]
  - [修复] 修复播放等按钮的色彩问题。
    - 有渐变色的情况下，这些按钮现在正常状态下为 G2 色，鼠标移入时为 G1 色。
    - 颜色转换时间现在是 0.25s。

v126d-pre3  更新时间：2020/10/10  [线下开发]
  - [修复] 修复缓存与“禁用主题色变换”之间的问题。
    - “禁用主题色变换”不再影响歌词文件编译结果。

v126d-pre2  更新时间：2020/10/10  [有严重问题] [线下开发]
  - [修改] 系统现在首选采用 csv 格式存储歌单。
    - 文件大小大幅缩小。
    - csv 与 json 的文件大小限制相同，因此 csv 可以存储更多歌曲（若评分为四位整数，则存储数量大约从 1039 提升至 2690）。
    - 歌单构造器上点击“保存”会以 csv 格式保存且别无选择。
    - 系统将尝试提醒用户重新保存其原有歌单，使其转换为 csv 格式。
    - 导入的 json 歌单仍以 json 方式存储。
    - 以 csv 存储的歌单在导出时会询问用户导出的格式。
    - 导出现在以窗口方式展现，而不是新标签页。
  - [修改] 歌单构造器导入数据的输入框改为多行代码输入框。
  - [修改] “以键入方式确认危险操作”的提示框现在允许其文字被选中。
  - [修复] 现在动态的“静态文件”允许浏览器缓存15天。
  - [修改] RemotePlay 歌曲现在具有随机生成（以歌曲、歌手名为种子）的主题色和渐变色。
  - [修复] 修复缓存导致 RemotePlay 出错的问题。
  - [修复] 修复 lrc 歌词切入时间为 0 的一句被删除的问题。
  - [添加][破坏] 添加渐变色的定义方式。
    - 必须手动修改 configuration.php，使得更新后不会发生错误。
    - configuration.php 中以常量定义形式存在，分别为 GC_COLOR_1 和 GC_COLOR_2。
    - 歌词文件的定义中在 [Meta] 段下，分别为 G1 和 G2。
    - 仅应用于顶栏以及歌单构造器的聚焦高亮。其他组件仍然使用默认主题色（MAIN_COLOR 或 A）

v126d-pre1  更新时间：2020/10/09  [线下开发]
  - [修改] 现在PC端上歌单构造器的歌曲列表宽度限制为 900px。
  - [修改] 歌单构造器操作高亮的背景色现在根据页面主题色着色。
    - 键盘焦点不在歌曲列表上时，显示为原来的灰色。
  - [修改] 现在歌单构造器操作高亮的背景色会透过聚焦的背景色显示出来。
  - [修改][破坏] 歌单编辑选择歌曲现在需要双击。
    - 单击会聚焦到该歌曲。
  - [添加] 添加了对转换后歌词进行文件系统缓存的功能（默认不开启）。
    - 将大约 120 首歌曲列表的处理时间从 1.5s 优化至大约 0.2s（线下开发环境）。
    - 详见 configuration.php。
    - 每个歌曲将额外消耗约 3~6 KB 的磁盘空间。
  - [修复] 非调试模式下不再用格式化方式输出歌词JSON。
  - [移除] 表面上移除 RojExplorer 歌单云保存功能。
    - 新的浏览器机制已导致其出现问题。
    - 原来已保存的歌单仍然可以《正常工作》，但是新的歌单不能再转换为此模式。
    - 《修复》了反复切换状态导致折叠按钮紊乱的问题。

v126c --更新时间：2020/10/08  [线下开发]
  - [添加] 歌单构造器现在支持键盘操作。
    - 可用按键有 WS、ED、A、R、ZXCV
  - [修复] 修复表示加载状态的模态窗口可以用 ESC 键关闭的问题。
    - 这是前端库的问题，难以直接修复。此处使用抛出错误的方法阻止窗口关闭。

v126b --更新时间：2020/10/08  [线下开发]
  - [修复] 移除了单击主页标题 6 次进入管理区的功能。
    - 管理区早已和用户系统合并。
  - [添加] “查看歌词配置”页面现在会输出两份标准歌词文件(lrc)以供通用的音乐播放软件或MP3设备使用。
  - [修改][破坏] 歌词文件中不再允许直接使用 HTML。它们会被转化为文本输出。
  - [修复] 移除播放队列。
    - 未来仍然计划添加该特性。
  - [修改] 由于模态窗口的引入，用户管理页面现在可以快速地修改用户的密码。
  - [修改] 顶栏现在显示用户名。
  - [修改] 需要多次确认的删除用户、删除歌单、导入歌单覆盖等危险行为现在以复述名称的方式来确认。
  - [修改] 所有 alert、alert2、prompt、confirm 相关内容均用模态窗口重写。
    - 有大量使用 async 与 await，使得老旧的浏览器不再受支持。
    - AmazeUI 的模态窗口类重新封装后被用于这些地方。重新封装后的模态窗口允许了键盘对窗口快捷的操作。
  - [修改] 角色颜色标签在激活时现在被着色。
  - [修复] 基本修复歌词缩放尺度较大造成的问题。
  - [修改] 单击歌词行跳转的功能现在仅在点中文字区域时有效。
  - [修改] 修改选中文字的样式。
  - [修复] 修复选中的文字无底色的问题。
  - [清理] 将 css 的着色部分（主题色变换部分，称为“着色剂”）从基础 css 中独立出来，以减小网络流量。
    - 例如：从 main.css.php 中独立出来的 css 为 main-colored.css.php。
  - [清理] 将 css 和 js 装入对应模块的文件夹。
  - [清理] 重构 function 中大多数函数的名称。
  - [修改] 歌单构造器上的保存方式窗口默认折叠。
  - [添加] 为歌单构造器添加导入导出功能。
  - [修改] 歌单构造器页面不再有固定刻（改进移动端体验）。
  - [修复] 歌单构造器页面现在显示歌单文件的实际大小。
  - [修复] 后端存储的用户歌单不再格式化。
  - [修复] 修复播放界面在较快网络下歌词初始位置不正确的问题。
  - [修复] 为歌词文件转换添加缓存机制。含有歌曲列表的页面加载时间大约变为原来的 1/6。
  - [修改] 歌单切换约束增加了一个限制条件。
  - [修复] 修复歌单构造器中的项目在添加后有时失去 ID 标签和评分标签点击事件的漏洞。
    - 这允许用户在对歌单进行删改后无需刷新就能毫无障碍地继续修改评分和 ID。
  - [修改][破坏] 配置文件改名为 configuration.php。
  - [修复] 允许软件在客户端无互联网的情况下完全正常运行（RemotePlay除外）。
  - [修复] 酷我音乐不再抓取歌曲主页，提高效率。
  - [修复] 酷我音乐爬虫不再认为标题为空的抓取结果是合法的。
  - [修改][破坏] 将酷我音乐 RemotePlay 编号的开头改为 'K_'
    - 'AK_' 开头的编号正常工作。
    - 歌单不再允许有 'AK_' 开头的编号。
  - [修改] 允许查看无音频歌曲的源代码
  - [修改] 为编译器添加部分前向兼容特性
    - 修改行注释机制。
      - 位于行首的“//”或“##”之后表示注释，直到此行结束。
      - 位于行中间的“##”之后表示注释，直到此行结束。但是如果“##”后面没有任何内容，则不认为这是注释。
    - 位于行首的 '!' 之后也被视为注释。
      - 即将引入新文件标准要求文件第一行标注其版本号（如“!dataver 201803”）。
  - [修改][破坏] 拆分了字体设置。
    - 设置页的提示未撰写。
    - “名称字体”“代码字体”暂时无效。
  - [修复][破坏] 对用户设置字体的注入漏洞使用了临时解决方案。
    - 如果设置项包含 !;<> 等字符之一，服务端接收保存设置请求时将阻止对该项进行修改。
    - 如果已有设置项包含不可接受的字符，系统会将其视为默认值。
    - 该特性未经过测试。
  - [修改][破坏] 用户设置中填写字体的项目现在不会自动加上引号，以允许正确地设置多个字体。
  - [修改] 将动态布局的延时调整为 1500ms。
    - 动态布局在加载之初会先执行一次。
  - [修复] 修复了大空格字体导致播放页布局错误的问题。
    - 修改了进度条动态布局数值。
  - [发布][破坏] 撤销 Gayhub 上的发布（重写后重新发布）
  - [修改] 对“新版样式”做了临终前最后的修改。
    - 大大改进了移动端性能。
    - 移除大部分动画。
  - [修复] 移除了右侧滑入菜单的动画，并且防止其在隐藏时渲染。
  - [修复] 优化了服务端性能。含有歌曲列表的页面加载时间大约变为原来的 1/6。
  - [修改] 查看歌词源代码现在会根据内容类型分页，并且包含了更多内容。
  - [添加] 现在编译歌词文件如果出现错误会给出错误位置。如果有不符合规范的写法会给出警告。
      - 酷我音乐RemotePlay生成的歌词文件不会被此忽略，并且通常会触发警告。
  - [清理] 拆分 tpl.function.php 并补充注释。
      - 这原来是一锅大杂烩
  - [修改] 将超级管理员变为一种用户类型。
      - 这移除了超级管理员的登入登出页面
      - 这废除了超级管理员明文密码的使用
      - 更新至此版本的用户需手动修改数据文件才能重新获得超级管理员访问权。全新安装不存在该问题。
  - [修复] 现在使用 [Comment] 插入注释不再会导致系统错误。
  - [修复] 修改默认字体 Noto Sans SC 的 webfont 链接至 jsDelivr
      - jsDelivr 中的webfont样式与字体文件从 fonts.googleapis.com 抓取。
      - 这用于解决 fonts.googleapis.com 在中国部分营业商/地区无法正常加载的问题。
  - [修改] 角色标签扩展至 1-7，并在编辑界面有了正确的标题。
      - 这允许在至多有3个歌手的歌曲中正确地区分演唱角色。
      - 参见演示站中 0403 号歌曲。
  - [修复] 修复点击歌词行有时跳到上一句歌词的漏洞。

configuration.php
<?php if(!defined('IN_SYSTEM')) exit;//Silence is golden ?><?php

define("MAIN_COLOR","4CAF50"); // [默认]主颜色
define("GC_COLOR_1",MAIN_COLOR);
define("GC_COLOR_2","CDDC39"); // 渐变色2

define("BASIC_URL",str_replace('__hostname__',$_SERVER['HTTP_HOST'],"http://__hostname__/software_build/txmp/"));

function _C(){
	return array(
		"app_name" => "阴乐播放器",//软件名称自定义
		"app_name_title" => "阴乐播放器（本地部署）",
		"timezone" => 0, //时区校准
		"cache_expire" => 24*60*60*93, //爬虫缓存时间（用于查看。爬虫较慢，建议缓存3个月）
		"temp_expire" => 3600, //歌词本缓存时间（用于下载。建议1小时）

		"can_register" => true, // 允许用户注册
		"ip_reg_limit" => 1, // 一个IP的注册限制量

		"user_playlist_quota" => 128, // 一个用户允许创建的最多歌单数量
		"user_exam_quota" => 50, // 允许用户创建的最大试卷数量
		"user_submission_capacity" => 20, // 保留的最近用户提交记录数量

		"exam_problem_limit" => 512, // 试卷中允许的最大的试题数量

		"offline_usage" => true, // 是否在离线环境下使用（将避免外源字体加载）

		"compiled_cache" => true, // 对已转换的歌词文件进行缓存
	);
}

date_default_timezone_set('Asia/Shanghai'); // 系统时区
